etna/analysis/plotters.py

```diff
@@ -264,62 +288,108 @@ def plot_backtest(
     history_df = df[df.index < forecast_start]
     backtest_df = df[df.index >= forecast_start]
+    freq_timedelta = df.index[1] - df.index[0]
+
+    # prepare colors
+    default_colors = plt.rcParams["axes.prop_cycle"].by_key()["color"]
+    color_cycle = itertools.cycle(default_colors)
+    lines_colors = {line_name: next(color_cycle) for line_name in ["history", "test", "forecast"]}
+
+    ax = prepare_axes(segments=segments, columns_num=columns_num, figsize=figsize)
     for i, segment in enumerate(segments):
         segment_backtest_df = backtest_df[segment]
         segment_history_df = history_df[segment]
+        segment_forecast_df = forecast_df[segment]
+        is_full_folds = set(segment_backtest_df.index) == set(segment_forecast_df.index)
 
-        if history_len:
-            plot_df = segment_history_df.tail(history_len)
+        # plot history
+        if history_len == "all":
+            plot_df = segment_history_df.append(segment_backtest_df)
+        elif history_len > 0:
+            plot_df = segment_history_df.tail(history_len).append(segment_backtest_df)
         else:
             plot_df = segment_backtest_df
+        ax[i].plot(plot_df.index, plot_df.target, color=lines_colors["history"])
 
-        ax[i].plot(plot_df.index, plot_df.target, label="history")
-        ax[i].plot(segment_backtest_df.index, segment_backtest_df.target, label="test")
-
-        segment_forecast_df = forecast_df[segment]
         for fold_number in folds:
-            forecast_df_slice_fold = segment_forecast_df[segment_forecast_df.fold_number == fold_number]
+            start_fold = fold_numbers[fold_numbers == fold_number].index.min()
+            end_fold = fold_numbers[fold_numbers == fold_number].index.max()
+            end_fold_exclusive = end_fold + freq_timedelta
+
+            # draw test
+            backtest_df_slice_fold = segment_backtest_df[start_fold:end_fold_exclusive]
+            ax[i].plot(backtest_df_slice_fold.index, backtest_df_slice_fold.target, color=lines_colors["test"])
+
+            if is_full_folds:
+                # draw forecast
+                forecast_df_slice_fold = segment_forecast_df[start_fold:end_fold_exclusive]
+                ax[i].plot(forecast_df_slice_fold.index, forecast_df_slice_fold.target, color=lines_colors["forecast"])
+            else:
+                forecast_df_slice_fold = segment_forecast_df[start_fold:end_fold]
+                backtest_df_slice_fold = backtest_df_slice_fold.loc[forecast_df_slice_fold.index]
+
+                # draw points on test
+                ax[i].scatter(backtest_df_slice_fold.index, backtest_df_slice_fold.target, color=lines_colors["test"])
+
+                # draw forecast
+                ax[i].scatter(
+                    forecast_df_slice_fold.index, forecast_df_slice_fold.target, color=lines_colors["forecast"]
+                )
+
+            # draw borders of current fold
+            opacity = 0.15 * ((fold_number + 1) % 2)
```

Comment:
Maybe opacity = 0.15 * ((fold_number + 1) % 2) + 0.15?
In plot_masks_partial_history when even folds are located between non-folds, the background does not change. it is hard to watch